pmd-apex/src/main/resources/category/apex/bestpractices.xml

```diff
@@ -106,6 +106,56 @@ trigger Accounts on Account (before insert, before update, before delete, after
         }
     }
 }
+]]>
+        </example>
+    </rule>
+    <rule name="ApexUnitTestAssertStatement"
+          since="6.12.0"
+          message="Apex test assert should have 2 params and assertEquals/assertNotEquals should have 3 params."
+          class="net.sourceforge.pmd.lang.apex.rule.bestpractices.ApexUnitTestAssertStatementRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_apex_bestpractices.html#apexunittestassertstatement">
+        <description>
+            Apex test assert should have 2 params and assertEquals/assertNotEquals should have 3 params.
+        </description>
+        <priority>3</priority>
+        <example>
+            <![CDATA[
+@isTest
+public class Foo {
+     @isTest
+    static void methodATest() {
+        System.assertNotEquals('123', o.StageName); // not good
+        System.assertEquals('123', o.StageName, o); // good
+        System.assert(1==1); // not good
+        System.assert(1==1, 2); // good
+    }
+}
+]]>
+        </example>
+    </rule>
+    <rule name="ApexUnitTestMethodShouldHaveIsTestAnnotation"
+          since="6.12.0"
+          message="Apex test methods should have @isTest annotation."
+          class="net.sourceforge.pmd.lang.apex.rule.bestpractices.ApexUnitTestMethodShouldHaveIsTestAnnotationRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_apex_bestpractices.html#apexunittestmethodshouldhaveistestannotation">
+        <description>
+            Apex test methods should have @isTest annotation either in the same line or in previous line.
```

Comment:
I'm not sure I understand the rule: Do all methods have to have the `@isTest` annotations? Or are there helper methods allowed in a test class?

> same line or previous line

Does that matter? At least in Java, the annotation always belongs to the following declaration regardless how much whitespace is in between. I don't know about Apex though...

Does this rule maybe want to solve the problem, that the `@isTest` method is maybe forgotten? Then this sounds like the new propose Java rule [DetachedTestCase](https://github.com/pmd/pmd/pull/1706).

Reply:
There is an existing PMD apex rule [ApexTestClassShouldHaveAssertsRule](https://pmd.github.io/latest/pmd_rules_apex_bestpractices.html#apexunittestclassshouldhaveasserts) that checks if a test class/method has an assert statement or not. This rule is kind of opposite to that. For example if there is a test method having assert statements then it should have `@isTest` annotation. (as [testMethod](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation_isTest.htm) has been deprecated and Salesforce advices to use `@isTest` annotation for test class/method, so I thought of adding this rule. However the helper methods in the test class need not have `@isTest` annotation.  
```
Does that matter?
```
No, It can have multiple new line spaces, but ideally there should not be more than one new line.